src/utils/UWebSocketTransport.ts

```diff
@@ -35,80 +43,159 @@ export async function createUWebSocketTransport<T extends { isConnected: boolean
-            // Send failed, socket might be closed
-            item.future.reject(new Error('Failed to send message: Socket closed'))
-            messageQueue.shift()
-            cleanup()
-            return
-          }
-          item.future.resolve()
+        const currentMessage = messageQueue[0]
+
+        // Check max retries
+        if (currentMessage.attempts >= maxRetryAttempts) {
+          logger.warn('Message dropped after max retries', {
+            transportId,
+            attempts: currentMessage.attempts,
+            messageSize: currentMessage.message.byteLength
+          })
+          currentMessage.future.reject(new Error('Message dropped after max retries'))
           messageQueue.shift()
-        } catch (error: any) {
-          item.future.reject(new Error(`Failed to send message: ${error.message}`))
-          messageQueue.shift()
-          cleanup()
-          return
+          continue
+        }
+
+        const result = processNextMessage(currentMessage)
+
+        if (result === UWebSocketSendResult.SUCCESS || result === UWebSocketSendResult.ERROR) {
+          continue
         }
-        await new Promise((resolve) => setTimeout(resolve, 0))
+
+        // Calculate exponential backoff delay
+        const backoffDelay = Math.min(retryDelayMs * Math.pow(2, currentMessage.attempts), maxBackoffDelayMs)
+
+        // Schedule retry with exponential backoff
+        processingTimeout = setTimeout(() => {
+          processingTimeout = null
+          void processQueue()
+        }, backoffDelay)
+
+        // Exit the loop after scheduling retry
+        return
       }
     } finally {
       isProcessing = false
+    }
+  }
 
-      if (messageQueue.length > 0) {
-        queueProcessingTimeout = setTimeout(() => {
-          void processQueue()
-        }, 1000)
+  function processNextMessage(item: QueuedMessage): UWebSocketSendResult {
+    try {
+      const result = socket.send(item.message, true)
+
+      // Only record metrics for known results
+      if (
+        result === UWebSocketSendResult.SUCCESS ||
+        result === UWebSocketSendResult.BACKPRESSURE ||
+        result === UWebSocketSendResult.DROPPED
+      ) {
+        metrics.observe(
+          'ws_message_size_bytes',
+          { result: UWebSocketSendResult[result].toLowerCase() },
+          item.message.byteLength
+        )
+      }
+
+      logger.debug('Processing queued message', {
+        transportId,
+        result,
+        queueLength: messageQueue.length,
+        messageAttempts: item.attempts,
+        messageSize: item.message.byteLength
+      })
+
+      switch (result) {
+        case UWebSocketSendResult.SUCCESS:
+          item.future.resolve()
+          messageQueue.shift()
+          break
+
+        case UWebSocketSendResult.BACKPRESSURE:
+          metrics.increment('ws_backpressure_events', { result: 'backpressure' })
+          item.attempts++ // Increment attempts to enable exponential backoff
```

Comment:
Following my previous message, these type of messages will be queued to be sent whenever possible, so we should move the retry process to the dropped ones if we want to give them a change to be sent, WDYT?